c2rust-analyze/src/rewrite/expr/mir_op.rs

```diff
@@ -461,6 +501,35 @@ impl<'a, 'tcx> ExprRewriteVisitor<'a, 'tcx> {
                         });
                     }
 
+                    Callee::IsNull => {
+                        self.enter_rvalue(|v| {
+                            let arg_lty = v.acx.type_of(&args[0]);
+                            if !v.flags[arg_lty.label].contains(FlagSet::FIXED) {
+                                let arg_non_null =
+                                    v.perms[arg_lty.label].contains(PermissionSet::NON_NULL);
+                                if arg_non_null {
+                                    v.emit(RewriteKind::IsNullToConstFalse);
+                                } else {
+                                    v.emit(RewriteKind::IsNullToIsNone);
+                                }
+                            }
+                        });
+                    }
+
+                    Callee::Null { .. } => {
+                        self.enter_rvalue(|v| {
+                            if !v.flags[pl_ty.label].contains(FlagSet::FIXED) {
+                                let arg_non_null =
+                                    v.perms[pl_ty.label].contains(PermissionSet::NON_NULL);
+                                if arg_non_null {
```

Comment:
Should it just be `assert!(!arg_non_null, "...");`?

Reply:
Fixed.  The `Null` case was originally symmetric with the `IsNull` case, but I don't think there's a good reason to have symmetry between the two.